src/util.h

```diff
@@ -221,6 +227,32 @@ static SECP256K1_INLINE void secp256k1_memczero(void *s, size_t len, int flag) {
     }
 }
 
+/* Cleanses memory to prevent leaking sensitive info. Won't be optimized out. */
+static SECP256K1_INLINE void secp256k1_memclear(void *ptr, size_t len) {
+#if defined(_MSC_VER)
+    /* SecureZeroMemory is guaranteed not to be optimized out by MSVC. */
+    SecureZeroMemory(ptr, len);
+#elif defined(__GNUC__)
+    /* We use a memory barrier that scares the compiler away from optimizing out the memset.
+     *
+     * Quoting Adam Langley <agl@google.com> in commit ad1907fe73334d6c696c8539646c21b11178f20f
+     * in BoringSSL (ISC License):
+     *    As best as we can tell, this is sufficient to break any optimisations that
+     *    might try to eliminate "superfluous" memsets.
+     * This method used in memzero_explicit() the Linux kernel, too. Its advantage is that it is
+     * pretty efficient, because the compiler can still implement the memset() efficently,
+     * just not remove it entirely. See "Dead Store Elimination (Still) Considered Harmful" by
+     * Yang et al. (USENIX Security 2017) for more background.
+     */
+    memset(ptr, 0, len);
+    __asm__ __volatile__("" : : "r"(ptr) : "memory");
+#else
+    void *(*volatile const volatile_memset)(void *, int, size_t) = memset;
+    volatile_memset(ptr, 0, len);
+#endif
+    SECP256K1_CHECKMEM_UNDEFINE(ptr, len);
```

Comment:
Hm, great point! The reason why `secp256k1_declassify` uses a run-time flag (instead of compile-time flag) is that we want to run the constant-time tests on the real binary.

I don't know what the performance impact of these additional instructions is, but I doubt that having a compile-time flag is a concern in this case. That means that we could just wrap the `SECP256K1_CHECKMEM_UNDEFINE` in an `#ifdef VERIFY`. @sipa What do you think?

Reply:
Oh, sorry, I think we're talking past each other. My suggestion, for which I'd like to have your opinion on, is to wrap `SECP256K1_CHECKMEM_UNDEFINE` in an `#ifdef VERIFY` block **here in this particular case of clearing memory**, i.e., inside `secp256k1_memclear` (and not in `secp256k1_declassify`.  Do you think that's a reasonable thing to do?

---
But regarding what you said:

> I think the reasoning is that we want the release binaries to be as close as possible to what we actually test in the ctime test, but disable the effect at runtime using `SECP256K1_CONTEXT_DECLASSIFY` to avoid a performance impact.

Yes, I agree. Perhaps an additional reason is to avoid any shenanigans that the valgrind syscalls may have. They should be noops, but it's certainly a bit safer not to run them at all. 

> It wouldn't surprise me that that is overkill; the conditional to determine whether or not to declassify operations at runtime may have a higher cost than actually executing the nop instructions that declassification actually correspond to when not running instrumented by valgrind.

I can imagine that the overhead of the noops is negligible, but at least checking the conditional should be negligible as well because we use `EXCEPT` to help the compiler predict the branch:
https://github.com/bitcoin-core/secp256k1/blob/68b55209f1ba3e6c0417789598f5f75649e9c14c/src/secp256k1.c#L236-L238

Reply:
I see. That's a valid point, which has been brought up before, see https://github.com/bitcoin-core/secp256k1/pull/813#issuecomment-691696004. (This thread is also very helpful to understand why we have run-time flag etc.) The conclusion was that the benefits outweigh the drawback that the build outputs depends on the availability of valgrind, but yeah, there's no perfect solution here. And yes, it's also auto-detected in autotools.

Reply:
Indeed. 

> That would mean `tests_noverify` wouldn't detect violations of use-after-clear

Fwiw, I had this downside in mind, and I don't think it's a big deal because we still have the normal tests.

> nor would applications linking against the production library when running in valgrind.

Okay, I had not considered this one... I still think that's acceptable, but yeah, it's a good point...